firedrake/interpolation.py

```diff
@@ -215,8 +250,58 @@ def __init__(
         self._allow_missing_dofs = allow_missing_dofs
         self.callable = None
 
+    @PETSc.Log.EventDecorator()
+    def interpolate(self, *function, output=None, transpose=False, default_missing_val=None):
+        """This performs interpolation by assembling `Interp`, which in turn calls
+        `Interpolator._interpolate`. Having this structure ensures consistency between
+        `Interpolator` and `Interp`.
+
+        This mechanism handles annotation since performing interpolation will drop an
+        Assemble block on the tape.
+
+        :arg function: If the expression being interpolated contains an
+            :class:`ufl.Argument`, then the :class:`.Function` value to
+            interpolate.
+        :kwarg output: Optional. A :class:`.Function` to contain the output.
+        :kwarg transpose: Set to true to apply the transpose (adjoint) of the
+              interpolation operator.
+        :returns: The resulting interpolated :class:`.Function`.
+        """
+        from firedrake.assemble import preprocess_base_form, assemble
+        from pyadjoint.tape import get_working_tape
+
+        V = self.V
+        if isinstance(V, firedrake.Function):
+            if not output:
+                # V can be the Function to interpolate into (e.g. see `Function.interpolate``).
+                output = V
+            V = V.function_space()
+        interp_data = {'subset': self.subset, 'freeze_expr': self.freeze_expr,
+                       'access': self.access, 'bcs': self.bcs,
+                       'allow_missing_dofs': self._allow_missing_dofs,
+                       'default_missing_val': default_missing_val}
+        interp = Interpolate(self.expr, V, interp_data=interp_data)
+        if transpose:
+            interp = adjoint(interp)
+
+        if function:
+            f, = function
+            # Passing in a function is equivalent to taking the action.
+            interp = action(interp, f)
+
+        get_working_tape()
+        # Assembly annotation records the form before it is preprocessed. Hence, we preprocess
+        # the form beforehand to facilitate derivation of the tangent linear/adjoint models.
+        # For example,
+        # -> `interp = Action(Interpolate(v1, v0), f)` with `v1` and `v0` being respectively `Argument`
+        # and `Coargument`. Differentiating `interp` is not currently supported as the action's left slot
+        # is a 2-form. However, after preprocessing, we obtain `Interpolate(f, v0)`, which can be differentiated.
+        interp = preprocess_base_form(interp)
+        res = assemble(interp, tensor=output, is_base_form_preprocessed=True)
+        return res
+
     @abc.abstractmethod
-    def interpolate(self, *args, **kwargs):
+    def _interpolate(self, *args, **kwargs):
```

Comment:
Could this have a more descriptive name? Looking at how it's used, perhaps `assemble` would be more appropriate?